sdk/modelsrepository/Azure.Iot.ModelsRepository/samples/ModelsRepositoryClientSamples/Program.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core.Diagnostics;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Azure.Iot.ModelsRepository.Samples
+{
+    public class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            using AzureEventSourceListener listener = AzureEventSourceListener.CreateTraceLogger(EventLevel.Verbose);
```

Comment:
What is this for? Should a customer do it to? Why would they want to do it? When?

Reply:
This construct will forward events from the client to System.Diagnostics.Trace.Writeline. One way to see it through the visual studio output window:


![image](https://user-images.githubusercontent.com/28658112/109047006-37489480-768a-11eb-8efb-7c5c832fbab9.png)

We have another construct in the `dmr-client` CLI that will forward client events to the console which is used in tandem with the `--debug` switch.

https://github.com/Azure/iot-plugandplay-models-tools/blob/dev/clients/dotnet/Azure.Iot.ModelsRepository.CLI/Program.cs#L41

Reply:
Creates a new instance of Azure.Core.Diagnostics.AzureEventSourceListener that forwards events to System.Diagnostics.Trace.WriteLine(System.Object).

Yes, if customers are interested in getting the events written as traces, they should create a listener.

Reply:
That would make a great code comment

Reply:
I changed the default listener and demonstrated how you can interact with the events by providing a EventWritten handler.

example:

```
[11:06:44:315][Informational] Client session 00390063-0035-0037-3600-650038003700 initialized with https content fetcher.
[11:06:44:378][Informational] Processing DTMI "dtmi:com:example:TemperatureController;1".
[11:06:44:381][Informational] Attempting to retrieve model content from "https://devicemodels.azure.com/dtmi/com/example/temperaturecontroller-1.json".
[11:06:44:403][Informational] Request [edda1706-1693-4960-aef3-5553a37cebe3] GET https://devicemodels.azure.com/dtmi/com/example/temperaturecontroller-1.json
x-ms-client-request-id:edda1706-1693-4960-aef3-5553a37cebe3
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Iot.ModelsRepository/1.0.0-alpha.20210224.1,(.NET Core 3.1.9; Microsoft Windows 10.0.18363)
client assembly: Azure.Iot.ModelsRepository
[11:06:44:802][Informational] Response [edda1706-1693-4960-aef3-5553a37cebe3] 200 OK (00.4s)
Accept-Ranges:REDACTED
Access-Control-Allow-Headers:REDACTED
Access-Control-Allow-Methods:REDACTED
Access-Control-Expose-Headers:REDACTED
Age:REDACTED
Date:Wed, 24 Feb 2021 19:06:42 GMT
ETag:"0x8D8A13DEABA3051"
Server:ECAcc,(sed/E17F)
X-Cache:REDACTED
x-ms-error-code:REDACTED
x-ms-request-id:REDACTED
x-ms-version:REDACTED
Content-MD5:REDACTED
Content-Type:application/json
Last-Modified:Tue, 15 Dec 2020 21:10:57 GMT
Content-Length:1762

[11:06:44:835][Informational] Discovered dependencies "dtmi:com:example:Thermostat;1", "dtmi:azure:DeviceManagement:DeviceInformation;1".
[11:06:44:837][Informational] Processing DTMI "dtmi:com:example:Thermostat;1".
[11:06:44:839][Informational] Attempting to retrieve model content from "https://devicemodels.azure.com/dtmi/com/example/thermostat-1.json".
[11:06:44:839][Informational] Request [563f1a36-ebed-4ee7-b391-9d388b2e21ce] GET https://devicemodels.azure.com/dtmi/com/example/thermostat-1.json
x-ms-client-request-id:563f1a36-ebed-4ee7-b391-9d388b2e21ce
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Iot.ModelsRepository/1.0.0-alpha.20210224.1,(.NET Core 3.1.9; Microsoft Windows 10.0.18363)
client assembly: Azure.Iot.ModelsRepository
[11:06:44:896][Informational] Response [563f1a36-ebed-4ee7-b391-9d388b2e21ce] 200 OK (00.1s)
Accept-Ranges:REDACTED
Access-Control-Allow-Headers:REDACTED
Access-Control-Allow-Methods:REDACTED
Access-Control-Expose-Headers:REDACTED
Age:REDACTED
Date:Wed, 24 Feb 2021 19:06:42 GMT
ETag:"0x8D8A13DEABC53BA"
Server:ECAcc,(sed/E15D)
X-Cache:REDACTED
x-ms-error-code:REDACTED
x-ms-request-id:REDACTED
x-ms-version:REDACTED
Content-MD5:REDACTED
Content-Type:application/json
Last-Modified:Tue, 15 Dec 2020 21:10:57 GMT
Content-Length:2469

[11:06:44:898][Informational] Processing DTMI "dtmi:azure:DeviceManagement:DeviceInformation;1".
[11:06:44:898][Informational] Attempting to retrieve model content from "https://devicemodels.azure.com/dtmi/azure/devicemanagement/deviceinformation-1.json".
[11:06:44:899][Informational] Request [9002dedc-4733-48ad-ba17-882912183566] GET https://devicemodels.azure.com/dtmi/azure/devicemanagement/deviceinformation-1.json
x-ms-client-request-id:9002dedc-4733-48ad-ba17-882912183566
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Iot.ModelsRepository/1.0.0-alpha.20210224.1,(.NET Core 3.1.9; Microsoft Windows 10.0.18363)
client assembly: Azure.Iot.ModelsRepository
[11:06:44:951][Informational] Response [9002dedc-4733-48ad-ba17-882912183566] 200 OK (00.1s)
Accept-Ranges:REDACTED
Access-Control-Allow-Headers:REDACTED
Access-Control-Allow-Methods:REDACTED
Access-Control-Expose-Headers:REDACTED
Age:REDACTED
Date:Wed, 24 Feb 2021 19:06:42 GMT
ETag:"0x8D8A13DEBC614F0"
Server:ECAcc,(sed/E110)
X-Cache:REDACTED
x-ms-error-code:REDACTED
x-ms-request-id:REDACTED
x-ms-version:REDACTED
Content-MD5:REDACTED
Content-Type:application/json
Last-Modified:Tue, 15 Dec 2020 21:10:59 GMT
Content-Length:2212

dtmi:com:example:TemperatureController;1 resolved in 3 interfaces.
[11:06:44:957][Informational] Client session 00300030-0062-0034-3500-310033003900 initialized with https content fetcher.
[11:06:44:957][Informational] Processing DTMI "dtmi:com:example:NotFound;1".
[11:06:44:958][Informational] Attempting to retrieve model content from "https://devicemodels.azure.com/dtmi/com/example/notfound-1.json".
[11:06:44:958][Informational] Request [527c37b7-1bcb-42ac-bf28-e48eba97da96] GET https://devicemodels.azure.com/dtmi/com/example/notfound-1.json
x-ms-client-request-id:527c37b7-1bcb-42ac-bf28-e48eba97da96
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Iot.ModelsRepository/1.0.0-alpha.20210224.1,(.NET Core 3.1.9; Microsoft Windows 10.0.18363)
client assembly: Azure.Iot.ModelsRepository
[11:06:45:055][Warning] Error response [527c37b7-1bcb-42ac-bf28-e48eba97da96] 404 Not Found (00.1s)
Access-Control-Allow-Headers:REDACTED
Access-Control-Allow-Methods:REDACTED
Access-Control-Expose-Headers:REDACTED
Date:Wed, 24 Feb 2021 19:06:42 GMT
Server:Windows-Azure-Web/1.0,Microsoft-HTTPAPI/2.0
x-ms-error-code:REDACTED
x-ms-request-id:REDACTED
x-ms-version:REDACTED
Content-Type:text/html
Content-Length:321

dtmi:com:example:NotFound;1 was not found in the default public models repository: Unable to resolve "dtmi:com:example:NotFound;1"'. Model file "https://devicemodels.azure.com/dtmi/com/example/notfound-1.json" not found or not accessible in target repository.

C:\Users\azad_\Enlistments\azure-sdk-for-net\artifacts\bin\ModelsRepositoryClientSamples\Release\netcoreapp3.1\ModelsRepositoryClientSamples.exe (process 4284) exited with code 0.
To automatically close the console when debugging stops, enable Tools->Options->Debugging->Automatically close the console when debugging stops.
```